homeassistant/components/api/__init__.py

```diff
@@ -385,6 +385,44 @@ async def post(
             )
 
         context = self.context(request)
+
+        if "return_response" in request.query or (
+            data is not None
+            and isinstance(data, dict)
+            and data.pop("return_response", False)
+        ):
+            if (
+                hass.services.supports_response(domain, service)
+                == ha.SupportsResponse.NONE
+            ):
+                return self.json_message(
+                    "Service does not support responses. Remove return_response from request.",
+                    HTTPStatus.BAD_REQUEST,
+                )
+
+            try:
+                # shield the service call from cancellation on connection drop
+                response = await shield(
+                    hass.services.async_call(
+                        domain,
+                        service,
+                        data,  # type: ignore[arg-type]
+                        blocking=True,
+                        context=context,
+                        return_response=True,
+                    )
+                )
+            except (vol.Invalid, ServiceNotFound) as ex:
+                raise HTTPBadRequest from ex
+
+            return self.json(response)
```

Comment:
This is going to have a completely different return type than if its not called with `return_response`.  I think this should be a different API

Reply:
Let's chat about this in the architecture repo.

Reply:
The architecture discussion is approved, it's OK to have a different shape of the data when returning service call response.